File: resume-backend/app/schemas/resume.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field


class BasicInfo(BaseModel):
    name: str
    phone: str
    email: str
    city: str


class JobPreference(BaseModel):
    target_role: str
    employment_type: str = ""
    expected_salary: str = ""


class EducationItem(BaseModel):
    school: str
    major: str = ""
    degree: str = ""
    start_date: str = ""
    end_date: str = ""


class EmploymentItem(BaseModel):
    company: str
    position: str = ""
    start_date: str = ""
    end_date: str = ""
    description: str = ""


class ProjectItem(BaseModel):
    name: str
    role: str = ""
    start_date: str = ""
    end_date: str = ""
    description: str = ""


class SkillCertificateInfo(BaseModel):
    skills: list[str] = Field(default_factory=list)
    certificates: list[str] = Field(default_factory=list)


class SectionVisibility(BaseModel):
    basic: bool = True
    job: bool = True
    education: bool = True
    employment: bool = True
    projects: bool = True
    skills: bool = True
    self_evaluation: bool = True


class ResumePayload(BaseModel):
    version: Literal[1] = 1
    basic: BasicInfo
    job: JobPreference
    education: list[EducationItem] = Field(default_factory=list)
    employment: list[EmploymentItem] = Field(default_factory=list)
    projects: list[ProjectItem] = Field(default_factory=list)
    skills: SkillCertificateInfo
    self_evaluation: str = ""
    section_visibility: SectionVisibility = Field(default_factory=SectionVisibility)

File: resume-miniprogram/src/types/resume.ts
export type TemplateId = "business" | "technology" | "graduate" | "analytics"

export interface JobIntelligence {
  version: 1
  roleName: string
  salaryByExperience: Record<string, string>
  responsibilities: string[]
  hardRequirements: string[]
  requiredSkills: string[]
  bonusSkills: string[]
  careerRoute: string[]
}

export interface ResumePayload {
  version: 1
  basic: { name: string; phone: string; email: string; city: string; gender: string }
  job: { targetRole: string; expectedSalary: string; availability: string }
  education: Array<{ school: string; major: string; degree: string; startDate: string; endDate: string; courses: string }>
  employment: Array<{ company: string; position: string; startDate: string; endDate: string; description: string }>
  projects: Array<{ name: string; role: string; startDate: string; endDate: string; description: string }>
  skills: { skills: string[]; certificates: string[]; englishLevel: string }
  selfEvaluation: string
  sectionVisibility: {
    basic: boolean; job: boolean; education: boolean; employment: boolean
    projects: boolean; skills: boolean; selfEvaluation: boolean
  }
}

export interface ResumeDraft {
  id?: string
  jobTitle: string
  templateId: TemplateId
  resume: ResumePayload
  jobIntelligence: JobIntelligence | null
}

export function createEmptyResume(): ResumePayload {
  return {
    version: 1,
    basic: { name: "", phone: "", email: "", city: "", gender: "" },
    job: { targetRole: "", expectedSalary: "", availability: "" },
    education: [],
    employment: [],
    projects: [],
    skills: { skills: [], certificates: [], englishLevel: "" },
    selfEvaluation: "",
    sectionVisibility: {
      basic: true, job: true, education: true, employment: true,
      projects: true, skills: true, selfEvaluation: true,
    },
  }
}

export function createEmptyDraft(): ResumeDraft {
  return {
    jobTitle: "",
    templateId: "business",
    resume: createEmptyResume(),
    jobIntelligence: null,
  }
}
